actors/eam/src/lib.rs

```diff
@@ -0,0 +1,289 @@
+            robust_address: exec4.robust_address,
+            eth_address,
+        }
+    }
+}
+
+#[derive(Serialize_tuple, Deserialize_tuple)]
+pub struct EvmConstructorParams {
+    /// The actor's "creator" (specified by the EAM).
+    pub creator: [u8; 20],
+    /// The initcode that will construct the new EVM actor.
+    pub initcode: RawBytes,
+}
+
+// fn eth2f4(addr: &[u8]) -> Result<Address, ActorError> {
+//     Address::new_delegated(EAM_ACTOR_ID, addr)
+//         .map_err(|e| ActorError::illegal_argument(e.to_string()))
+// }
+
+fn assert_code_size(code: &[u8]) -> Result<(), ActorError> {
+    (code.len() == MAX_CODE_SIZE).then(|| ()).ok_or(ActorError::illegal_argument(
+        "Supplied EVM bytecode is larger than 24kB.".to_string(),
+    ))
+}
+
+/// hash of data with Keccack256, with first 12 bytes cropped
+fn hash_20<BS, RT>(rt: &RT, data: &[u8]) -> [u8; 20]
+where
+    BS: Blockstore + Clone,
+    RT: Runtime<BS>,
+{
+    let buf = rt.hash(SupportedHashes::Keccak256, data);
+    buf[12..32].try_into().unwrap()
+}
+
+fn create_actor<BS, RT>(
+    rt: &mut RT,
+    creator: [u8; 20],
+    new_addr: [u8; 20],
+    initcode: Vec<u8>,
+) -> Result<RawBytes, ActorError>
+where
+    BS: Blockstore + Clone,
+    RT: Runtime<BS>,
+{
+    let constructor_params =
+        RawBytes::serialize(EvmConstructorParams { creator, initcode: initcode.into() })?;
+
+    let init_params = Exec4Params {
+        code_cid: rt.get_code_cid_for_type(Type::EVM),
+        constructor_params,
+        subaddress: new_addr.to_vec().into(),
+    };
+
+    let ret: ext::init::Exec4Return = rt
+        .send(
+            &INIT_ACTOR_ADDR,
+            ext::init::EXEC4_METHOD,
+            RawBytes::serialize(&init_params)?,
+            rt.message().value_received(),
+        )?
+        .deserialize()?;
+
+    Ok(RawBytes::serialize(EamReturn::from_exec4(ret, new_addr))?)
+}
+
+/// lookup caller's raw ETH address
+fn get_eth_address<BS, RT>(rt: &RT) -> Result<[u8; 20], ActorError>
+where
+    BS: Blockstore + Clone,
+    RT: Runtime<BS>,
+{
+    let caller_id = rt.message().caller().id().unwrap();
+
+    let addr = rt.lookup_address(caller_id);
+
+    match addr.map(|a| *a.payload()) {
+        Some(Payload::Delegated(eth)) => Ok(eth.subaddress().try_into().unwrap()),
+        _ => Err(ActorError::assertion_failed(
+            "All FEVM actors should have a delegated address".to_string(),
+        )),
+    }
+}
+
+pub struct EamActor;
+impl EamActor {
+    pub fn constructor<BS, RT>(rt: &mut RT) -> Result<(), ActorError>
+    where
+        BS: Blockstore + Clone,
+        RT: Runtime<BS>,
+    {
+        let actor_id = rt.resolve_address(&rt.message().receiver()).unwrap();
+        if actor_id != EAM_ACTOR_ID {
+            return Err(ActorError::forbidden(format!(
+                "The Ethereum Address Manager must be deployed at {EAM_ACTOR_ID}, was deployed at {actor_id}"
+            )));
+        }
+        rt.validate_immediate_caller_accept_any()
```

Comment:
Hm. I should have checked `rt.validate_immediate_caller_is(std::iter::once(&INIT_ACTOR_ADDR))?;`